src/fetch.rs

```diff
@@ -317,3 +253,84 @@ fn get_cargo_toml_from_git_url(url: &str) -> Result<String> {
         .chain_err(|| "Git response not a valid `String`")?;
     Ok(body)
 }
+
+/// Generate all similar crate names
+///
+/// Examples:
+///
+/// | input | output |
+/// | ----- | ------ |
+/// | cargo | cargo  |
+/// | cargo-edit | cargo-edit, cargo_edit |
+/// | parking_lot_core | parking_lot_core, parking_lot-core, parking-lot_core, parking-lot-core |
+fn gen_fuzzy_crate_names(crate_name: String) -> Result<Vec<String>> {
+    const PATTERN: [u8; 2] = [b'-', b'_'];
+
+    let wildcard_indexs = crate_name
+        .bytes()
+        .enumerate()
+        .filter(|(_, item)| PATTERN.contains(item))
+        .map(|(index, _)| index)
+        .collect::<Vec<usize>>();
+    if wildcard_indexs.len() > 126 {
+        return Err(ErrorKind::TooManyWildcardsInCrateName.into());
+    } else if wildcard_indexs.is_empty() {
+        return Ok(vec![crate_name]);
+    }
+
+    let mut result = vec![];
+    let mut bytes = crate_name.into_bytes();
+    for mask in 0..2u128.pow(wildcard_indexs.len() as u32) {
+        for (mask_index, wildcard_index) in wildcard_indexs.iter().enumerate() {
+            let mask_value = (mask >> mask_index) & 1 == 1;
+            if mask_value {
+                bytes[*wildcard_index] = b'-';
+            } else {
+                bytes[*wildcard_index] = b'_';
+            }
+        }
+        result.push(String::from_utf8(bytes.clone()).unwrap());
+    }
+    Ok(result)
+}
+
+#[test]
+fn test_gen_fuzzy_crate_names() {
+    fn test_helper(input: &str, expect: &[&str]) {
+        let mut actual = gen_fuzzy_crate_names(input.to_string()).unwrap();
+        actual.sort();
+
+        let mut expect = expect.iter().map(|x| x.to_string()).collect::<Vec<_>>();
+        expect.sort();
+
+        assert_eq!(actual, expect);
+    }
+
+    test_helper("", &[""]);
+    test_helper("-", &["_", "-"]);
+    test_helper("DCjanus", &["DCjanus"]);
+    test_helper("DC-janus", &["DC-janus", "DC_janus"]);
+    test_helper(
+        "DC-_janus",
+        &["DC__janus", "DC_-janus", "DC-_janus", "DC--janus"],
+    );
+}
+
+fn summary_raw_path(crate_name: &str) -> String {
+    match crate_name.len() {
+        0 => unreachable!(),
```

Comment:
where do we check the `crate_name` is not empty? I would appreciate a comment inside of `unreachable` like
```rust
unreachable!("we check that crate_name is not empty here");
```